### ISSUE TITLE
Deploy multiple functions codebases from root

### ISSUE THREAD

**Issue by zallesov** [0x80/isolate-package#100]
Hello. First, I'd like to show my appreciation for building this project. This topic has been a pain in the butt for years.
I used to use firelink which works too. But the solution you've provided is much more solid. I hope it will be merged upstream one day.

In my project, there are several packages with cloud functions separated by what are called codebases. They are deployed separately. I have .firebase.json and .firebaserc in the root of the project in which the codebases are defined.

```
├─ functions1
   └── packaje.json
├─ functions2
   └── packaje.json
│ .firebaserc
│ firebase.json
│ isolate.config.json
```
https://github.com/0x80/isolate-package/blob/main/docs/firebase.md#firebase
From this documentation, it seems like it is impossible to set up a deployment from the root of multiple function codebases.
`targetPackagePath` in isolate.config.json can only point to one codebase that can use isolation.

Can you confirm that?

The workaround would be to set up .firebaserc and firebase.json for each package. right?

**Comment by 0x80** [comment]
Yet it's currently assumed that you deploy packages separately. I will consider supporting having a firebase.json in the root, because it could be essential if you want to run all packages simultaneously with the emulators, but currently I test each package in isolation. 

Could you let me know what your firebase.json looks like? I'm not familiar with defining multiple codebases in a single json file.

There's a related issue here https://github.com/0x80/firebase-tools-with-isolate/issues/8

**Comment by zallesov** [comment]
Thanx for the quick reaction.
Here is how one can set up multiple codebases in one firebase.json
```
{
  ...
  "functions": [
    {
      "source": "functions",
      "codebase": "default",
      "ignore": [
        "node_modules",
        ".git",
        "firebase-debug.log",
        "firebase-debug.*.log"
      ],
      "predeploy": [
        "pnpm --prefix \"$RESOURCE_DIR\" run build"
      ],
      "runtime": "nodejs20",
      "isolate": true
    },
    {
      "source": "sound-attacher/isolate",
      "codebase": "sound-attacher",
      "ignore": [
        "node_modules",
        ".git",
        "firebase-debug.log",
        "firebase-debug.*.log"
      ],
      "predeploy": [
        "pnpm run sound-attacher:build",
        "isolate"
      ],
      "runtime": "nodejs20",
      "isolate": true
    }],
  "emulators": {
    "hosting": {
      "host": "localhost",
      "port": "5002",
      "region": "europe-west1"
    }
  }
}

````
Docs https://firebase.google.com/docs/functions/organize-functions?gen=2nd#managing_multiple_source_packages_monorepo

I invision the solution can be also to make `targetPackagePath` a list of some sort pointing to the codebases roots.

**Comment by 0x80** [comment]
I haven't read the firebase docs yet, but I'm wondering...

Would firebase get confused if you, in addition to having this file, also create separate firebase.json files in each package?

Maybe you can use this configuration for running the emulators, but by placing a firebase.json in each package (without emulator config), you can then isolate/deploy each package separately using the current functionality.

Note that it might be necessary probably to use firebase-tools-with-isolate, because you do not want your emulators to point to the isolate directory I think. It breaks hot loading.

**Comment by zallesov** [comment]
Thanx you for the response. I'm currently implementing exactly what you suggest. Except I now only have partial firebase.json and .firebaserc for each subproject. not ideal. each function source has to run on a separate port in its own emulator instance

**Comment by 0x80** [comment]
@zallesov yes that's how I do it, but I didn't have many packages that deploy to functions. Each package for me deploys many cloud functions.

I hope I find some time to look into supporting the root firebase.json, because I like the idea of simplifying the emulator setup, but I don't expect it to be soon. I'm very much behind on the backlog of my paid projects at the moment.

I must say the way Firebase suggests to structure a monorepo is pretty dubious, as it doesn't take into account that a monorepo typically also contains shared packages and other deployment platforms. I would always expect all packages to be in a shared subfolder, so I'm hoping the firebase.json definition would allow that.

**Comment by phucledien** [comment]
This is crucial if you wanna deploy all the cloud func at once 🤔 . 
In my project, I have to cd to each package (fn) to deploy each of them. which quite time consuming since can't deploy in parallel (which firebase support by default when use npx firebase deploy).
Have try writing script and use the `&` to run each step in parallel, but seems not really reliable, since will got random port conflict error from firebase tool 😢

**Comment by 0x80** [comment]
Happy to report that this is now supported. See the announcement [here](https://github.com/firebase/firebase-tools/issues/653#issuecomment-2585895455)
